app/services/product_tower_service.py

```
@@ -0,0 +1,304 @@
+            if user_norm == 0 or product_norm == 0:
+                return 0.0
+            
+            similarity = dot_product / (user_norm * product_norm)
+            return float(np.clip(similarity, 0.0, 1.0))
+            
+        except Exception as e:
+            logger.error(f"유사도 계산 오류: {e}")
+            return 0.0
+    
+    def generate_recommendation_reason(self, product: Product, 
+                                     matched_keywords: List[str],
```

Flake8 (7.2.0)

[error] continuation line under-indented for visual indent

(E128)